Unable to scroll through Notes, Mail, and Photos on iCloud.com using external mouse

lilyspiniolas opened the pull request:
#### fb629be740442d05db62a5f2fcb98e867500804a
<pre>
Unable to scroll through Notes, Mail, and Photos on iCloud.com using external mouse
<a href="https://bugs.webkit.org/show_bug.cgi?id=286200">https://bugs.webkit.org/show_bug.cgi?id=286200</a>
<a href="https://rdar.apple.com/136683606">rdar://136683606</a>

Reviewed by Tim Horton and Abrar Rahman Protyasha.

Prior to these changes, iOS did not pass wheel events to subframes,
which resulted in scrolling not working on iCloud.com.
The WK2 codepath in &apos;EventHandler::passWheelEventToWidget&apos; for iOS
now behaves the same as its Mac counterpart, resulting in wheel
events being dispatched to subframes appropriately. The function&apos;s
other definitions were also updated to use RefPtrs instead of
raw pointers in the WK2 codepath.

To test these changes, &apos;WheelEventDispatchedToSubframe&apos;
was added to WKScrollViewTests. The test sets a global
variable, which is updated by a wheel event listener
set by the iframe. A scroll is performed on the web
view, and if the global variable gets updated,
the test passes.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::passWheelEventToWidget):
* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(TEST(WKScrollViewTests, WheelEventDispatchedToSubframe)):

Canonical link: <a href="https://commits.webkit.org/289256@main">https://commits.webkit.org/289256@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27af94997e256c9665e61181567a328ec4c6d760

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86052 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5671 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40418 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91063 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36958 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88097 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5910 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/13738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66764 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/24565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89055 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4523 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78035 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47072 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4356 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32336 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36045 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33194 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/92877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13299 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9723 "") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75531 "77 flakes 102 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13513 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73896 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74700 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18920 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17328 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13398 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13329 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/18684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13104 "Failed to compile WebKit") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16562 "Failed to compile WebKit") | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14891 "Failed to compile WebKit") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -412,14 +412,18 @@ static bool findViewInSubviews(NSView *superview, NSView *target)
-        // WK2 code path. No wheel events on iOS anyway.
-        return false;
+        // WebKit2 code path.
+        auto* frameView = dynamicDowncast<LocalFrameView>(widget);
```
```suggestion
        RefPtr frameView = dynamicDowncast<LocalFrameView>(widget);
```

hortont424 commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -412,14 +412,18 @@ static bool findViewInSubviews(NSView *superview, NSView *target)
 }
 
-bool EventHandler::passWheelEventToWidget(const PlatformWheelEvent&, Widget& widget, OptionSet<WheelEventProcessingSteps>)
+bool EventHandler::passWheelEventToWidget(const PlatformWheelEvent& wheelEvent, Widget& widget, OptionSet<WheelEventProcessingSteps> processingSteps)
```
Can we merge these into an EventHandlerCocoa method? They seem pretty close.

hortont424 replied:
Maybe not close enough. Not for this patch, anyway

hortont424 commented on Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
```diff
@@ -498,4 +498,28 @@ static void traverseLayerTree(CALayer *layer, void(^block)(CALayer *))
+TEST(WKScrollViewTests, WheelEventDispatchedToSubframe)
+{
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 200, 200)]);
+    [webView synchronouslyLoadHTMLString:@"<!DOCTYPE html>"
```
maybe a raw string literal like the PSON tests?

aprotyas commented on Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
```diff
@@ -498,4 +498,28 @@ static void traverseLayerTree(CALayer *layer, void(^block)(CALayer *))
+    EXPECT_TRUE([[webView objectByEvaluatingJavaScript:@"window.subframeHit"] intValue]);
+#endif
+    [webView synchronouslyHandleScrollEventWithPhase:WKBEScrollViewScrollUpdatePhaseBegan location:CGPointMake(100, 100) delta:CGVectorMake(0, 10)];
+    EXPECT_TRUE([[webView objectByEvaluatingJavaScript:@"window.subframeHit"] intValue]);
```
This will do duplicate work for the `!USE(BROWSERENGINEKIT)` case, no?

aprotyas replied:
```suggestion
#if !USE(BROWSERENGINEKIT)
    [webView synchronouslyHandleScrollEventWithPhase:UIScrollPhaseMayBegin location:CGPointMake(100, 100) delta:CGVectorMake(0, 0)];
#else
    [webView synchronouslyHandleScrollEventWithPhase:WKBEScrollViewScrollUpdatePhaseBegan location:CGPointMake(100, 100) delta:CGVectorMake(0, 10)];
#endif
    EXPECT_TRUE([[webView objectByEvaluatingJavaScript:@"window.subframeHit"] intValue]);
```

aprotyas replied:
Also should the `delta:` argument values not be identical in both cases?

aprotyas commented on Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
```diff
@@ -498,4 +498,31 @@ static void traverseLayerTree(CALayer *layer, void(^block)(CALayer *))
+#else
+    [webView synchronouslyHandleScrollEventWithPhase:WKBEScrollViewScrollUpdatePhaseBegan location:CGPointMake(100, 100) delta:CGVectorMake(0, 10)];
+    EXPECT_TRUE([[webView objectByEvaluatingJavaScript:@"window.subframeHit"] intValue]);
+#endif
```
Totally optional, but you can write the expectation once outside of the conditionals!

```suggestion
#if !USE(BROWSERENGINEKIT)
    [webView synchronouslyHandleScrollEventWithPhase:UIScrollPhaseMayBegin location:CGPointMake(100, 100) delta:CGVectorMake(0, 10)];
#else
    [webView synchronouslyHandleScrollEventWithPhase:WKBEScrollViewScrollUpdatePhaseBegan location:CGPointMake(100, 100) delta:CGVectorMake(0, 10)];
#endif
    EXPECT_TRUE([[webView objectByEvaluatingJavaScript:@"window.subframeHit"] intValue]);
```

lilyspiniolas committed:
Unable to scroll through Notes, Mail, and Photos on iCloud.com using external mouse
https://bugs.webkit.org/show_bug.cgi?id=286200
rdar://136683606

Reviewed by Tim Horton and Abrar Rahman Protyasha.

Prior to these changes, iOS did not pass wheel events to subframes,
which resulted in scrolling not working on iCloud.com.
The WK2 codepath in 'EventHandler::passWheelEventToWidget' for iOS
now behaves the same as its Mac counterpart, resulting in wheel
events being dispatched to subframes appropriately. The function's
other definitions were also updated to use RefPtrs instead of
raw pointers in the WK2 codepath.

To test these changes, 'WheelEventDispatchedToSubframe'
was added to WKScrollViewTests. The test sets a global
variable, which is updated by a wheel event listener
set by the iframe. A scroll is performed on the web
view, and if the global variable gets updated,
the test passes.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::passWheelEventToWidget):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::passWheelEventToWidget):
* Tools/TestWebKitAPI/Tests/ios/WKScrollViewTests.mm:
(TEST(WKScrollViewTests, WheelEventDispatchedToSubframe)):

Canonical link: https://commits.webkit.org/289256@main
